atom/src/member.cpp

```diff
@@ -960,27 +997,43 @@ struct RemoveTask : public BaseTask
 
 } // namespace
 
+bool Member::has_observers( uint8_t change_types )
+{
+    if ( static_observers ) {
+        std::vector<Observer>::iterator it;
+        std::vector<Observer>::iterator end = static_observers->end();
+        for( it = static_observers->begin(); it != end; ++it )
+        {
+            if( it->enabled( change_types ) )
+                return true;
+        }
+    }
+    return false;
+}
 
 void
-Member::add_observer( PyObject* observer )
+Member::add_observer( PyObject* observer, uint8_t change_types )
 {
     if( modify_guard )
     {
-        ModifyTask* task = new AddTask( this, observer );
+        ModifyTask* task = new AddTask( this, observer, change_types );
         modify_guard->add_task( task );
         return;
     }
     if( !static_observers )
-        static_observers = new std::vector<cppy::ptr>();
+        static_observers = new std::vector<Observer>();
     cppy::ptr obptr( cppy::incref( observer ) );
-    std::vector<cppy::ptr>::iterator it;
-    std::vector<cppy::ptr>::iterator end = static_observers->end();
+    std::vector<Observer>::iterator it;
+    std::vector<Observer>::iterator end = static_observers->end();
     for( it = static_observers->begin(); it != end; ++it )
     {
-        if( *it == obptr || utils::safe_richcompare( it->get(), obptr, Py_EQ ) )
+        if( it->match( obptr ) ) {
+            // TODO: Should it update change types here?
+            it->m_change_types = change_types;
```

Comment:
This is a bit tricky. I am inclined to say we should OR the flags, but then the first registered callback may get unexpected notifications and when removing observer we would need to undo the flag. I think the case is rare enough that we could simply error. But I am not sure how to propagate the error...

Reply:
The comment was about whether re-adding should update the change types vs adding another function to do that. I think just updating is the most simple since it checks anyways.

Reply:
Previously when adding multiple times the same observer we were simply making we actually connected it a single time. What bothers me here is that code registering the same handler with different change types will race against one another, which is why I would prefer to error if the change types do not match.

Reply:
It also updates the change type when add is called in the observer pool.

Reply:
I don't see your point. The pool is for dynamic observer and I would make the same argument there.

Reply:
Was just mentioning it so whatever change needs made is done in both places.  It looks like the insert there was not correct.

Reply:
Can you give an example of the problematic case(s) you had in mind?

With how it is now I think there should still only ever be one observer per python object? I tried to keep it that way but maybe I messed something up unintentionally, I don't have much c++ experience.

Reply:
One contrived example that popped up in an example written by @clamdad in #158 is:

```python
class InventoryItem(Atom):

    @observe("unit_price")
    def _observe_unit_price(self, change):
        savings = change.get("oldvalue", 0) - change.get("value")
        if savings > 0:
            print(f"Save ${savings} now on {self.name}s!")
```

It is a weird case where one requests an observer on unit_price in 2 ways (through a magic name and a decorator). But if now one writes:

```python
class InventoryItem(Atom):

    @observe("unit_price", ChangeType.CREATE)
    def _observe_unit_price(self, change):
        savings = change.get("oldvalue", 0) - change.get("value")
        if savings > 0:
            print(f"Save ${savings} now on {self.name}s!")
```

now the two declaration are not equivalent and it is far from obvious which one should/will win. This is why I would prefer to simply error in such a case. Since this is the only way add_observer can fail going from void to bool would be sufficient to propagate the error I think. Does that make sense ?

Reply:
I did not consider the ModifyGuard I will need to review this more carefully.